# 🔢 Notebook 04: Embeddings and Vector Representations

**LangChain 1.0.5+ | Mixed Level Class**

## 🎯 Objectives
1. Understand what embeddings are
2. Use OpenAI Embeddings
3. Use Google Gemini Embeddings
4. Compare embedding models
5. Calculate similarity between vectors

In [8]:
from dotenv import load_dotenv
import os
load_dotenv()
print("✅ Setup complete")

✅ Setup complete


## 1. What are Embeddings? 🤔

### 🔰 BEGINNER

**Embeddings** convert text into numbers (vectors) that capture meaning.

Think of it like a GPS coordinate:
- "dog" → [0.2, 0.8, 0.1, ...] (1536 numbers)
- "cat" → [0.3, 0.7, 0.2, ...] (close to "dog"!)
- "car" → [0.9, 0.1, 0.8, ...] (far from "dog")

Similar meanings = Similar vectors!

## 2. OpenAI Embeddings

In [3]:
# from langchain_openai import OpenAIEmbeddings
from langchain_ollama import OllamaEmbeddings

# Initialize
# embeddings = OpenAIEmbeddings(model="text-embedding-3-small")
embeddings = OllamaEmbeddings(model="embeddinggemma:latest")

# Create embedding for a query
query = "What is machine learning?"
query_vector = embeddings.embed_query(query)

print(f"Query: {query}")
print(f"Vector dimensions: {len(query_vector)}")
print(f"First 5 values: {query_vector[:5]}")

# Embed multiple documents
docs = [
    "Machine learning is a subset of AI",
    "Deep learning uses neural networks",
    "The weather is sunny today"
]

doc_vectors = embeddings.embed_documents(docs)
print(f"\nEmbedded {len(doc_vectors)} documents")

Query: What is machine learning?
Vector dimensions: 768
First 5 values: [-0.06847741, -0.040785395, -0.01014158, 0.023063164, 0.0341972]

Embedded 3 documents


## 3. Google Gemini Embeddings

In [4]:
# !uv pip install langchain-google-genai
!pip install langchain-google-genai


[notice] A new release of pip is available: 26.2 -> 26.2.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [9]:
# Google Gemini Embeddings
if os.getenv("GOOGLE_API_KEY"):
    from langchain_google_genai import GoogleGenerativeAIEmbeddings
    
    gemini_embeddings = GoogleGenerativeAIEmbeddings(model="models/gemini-embedding-001")
    
    vector = gemini_embeddings.embed_query("Hello world")
    print(f"Gemini vector dimensions: {len(vector)}")
else:
    print("⚠️ Google API key not found. Set GOOGLE_API_KEY in .env")

Both GOOGLE_API_KEY and GEMINI_API_KEY are set. Using GOOGLE_API_KEY.


Gemini vector dimensions: 3072


## 4. Similarity Calculation

In [10]:
import numpy as np

def cosine_similarity(vec1, vec2):
    """Calculate cosine similarity between two vectors"""
    return np.dot(vec1, vec2) / (np.linalg.norm(vec1) * np.linalg.norm(vec2))

# Test similarity
texts = [
    "I love programming",
    "I enjoy coding",
    "The weather is nice"
]

vectors = [embeddings.embed_query(text) for text in texts]

print("Similarity Scores:")
print(f"'{texts[0]}' vs '{texts[1]}': {cosine_similarity(vectors[0], vectors[1]):.4f}")
print(f"'{texts[0]}' vs '{texts[2]}': {cosine_similarity(vectors[0], vectors[2]):.4f}")
print("\nSimilar meanings = Higher similarity score!")

Similarity Scores:
'I love programming' vs 'I enjoy coding': 0.8476
'I love programming' vs 'The weather is nice': 0.4585

Similar meanings = Higher similarity score!


## Summary

✅ Embeddings convert text to vectors
✅ OpenAI text-embedding-3-small: 1536 dimensions
✅ Google Gemini: 768 dimensions
✅ Cosine similarity measures vector similarity

**Next:** Vector Stores for efficient search!